torch/testing/_internal/distributed/distributed_test.py

```diff
@@ -5123,52 +5123,67 @@ def test_output_unused_in_loss(self):
             else:
                 # It is permissible for other ranks that did not fail to
                 # successfully exit or crash in the monitored barrier, the main
                 # purpose of monitored barrier is to report the rank that hung
                 # on rank 0.
                 try:
-                    dist.monitored_barrier(None, timeout)
-                except RuntimeError as e:
+                    dist.monitored_barrier(timeout=timeout)
+                except RuntimeError:
                     pass
 
-        @require_backend({"gloo"})
-        @require_backends_available({"gloo"})
-        def test_monitored_barrier_gloo_subgroup(self):
-            failed_rank = 1
-            timeout = 5
-            subgroup = dist.new_group(ranks=[0, 1])
-            if self.rank == failed_rank:
+        @require_backend({"gloo", "nccl"})
+        @require_backends_available({"gloo", "nccl"})
+        def test_monitored_barrier_allreduce_hang(self):
+            # tests expected behavior when nonzero rank hangs.
+            if "NCCL_ASYNC_ERROR_HANDLING" in os.environ:
+                del os.environ["NCCL_ASYNC_ERROR_HANDLING"]
+            os.environ["NCCL_BLOCKING_WAIT"] = "1"
+            nccl_pg = dist.new_group(
+                ranks=list(i for i in range(int(self.world_size))),
+                timeout=timedelta(seconds=10),
+                backend=dist.Backend.NCCL,
+            )
+            gloo_pg = dist.new_group(
+                ranks=list(i for i in range(int(self.world_size))),
+                backend=dist.Backend.GLOO,
+            )
+            tensors = [
+                torch.ones(10, device=self.rank) * self.rank
+            ]
+            # Let all ranks call allreduce first to set up communicators etc.
+            # Directly simulating error here will run into store issue described
+            # in https://github.com/pytorch/pytorch/issues/54524.
+            nccl_pg.allreduce(tensors).wait()
+            # All ranks besides 0 call into allreduce.
+            if self.rank != 0:
+                with self.assertRaisesRegex(RuntimeError, "Caught collective operation timeout"):
+                    nccl_pg.allreduce(tensors).wait()
                 return
-            if self.rank == 0:
-                with self.assertRaisesRegex(RuntimeError, f"Rank {failed_rank}"):
-                    dist.monitored_barrier(subgroup, timeout)
+
+            # Rank 0 should report timed out rank.
+            monitored_barrier_timeout_seconds = timedelta(seconds=2)
+            world_size = int(self.world_size)
+            if world_size == 2:
+                err_regex = "Rank 1"
             else:
-                # Other ranks call into monitored barrier, but should not erorr
-                # out since the barrier is a noop because they are not part of
-                # the subgroup
-                dist.monitored_barrier(subgroup, timeout)
+                err_regex = f"Rank [1-{world_size - 1}]"
```

Comment:
What exactly is this matching? For world size of 4, this would be "rank [1-3]" but looking at https://github.com/pytorch/pytorch/pull/53773, I don't see any such string in the error we raise? Should the error regex also contain "failed to pass monitoredBarrier"?

Reply:
https://github.com/pytorch/pytorch/pull/53773/files#diff-fca17800a316f138000a4cd30902ac0a5fd2b4cd904b562d94e465e5d8255da3R2731 should report "rank x failed to cross barrier" where x is a rank that failed to cross the barrier.

Don't want to put too many constraints on the implementation, so the test just checks that it reported the failure for rank 1, or rank 2, or rank 3 for example, so the regex is rank [1-3]. This is due to iterating the unordered_map, the error can be raised for any rank that did not cross the barrier. Planning to add functionality to report all failed ranks in a follow up change.